File: storage/ndb/nodejs/test/integraltypes/QueryTableScanTest.js
/*
 Copyright (c) 2013, Oracle and/or its affiliates. All rights
 reserved.
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; version 2 of
 the License.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 02110-1301  USA
 */
"use strict";

var udebug = unified_debug.getLogger("integraltypes/QueryTableScan.js");

var t0 = new harness.ConcurrentTest('test');
t0.run = function() {
  this.pass();
};

var q1 = {p1: 4, p2: 6, expected: [4, 5, 6], queryType: 3, ordered: false, predicate: function(qdt) {
  return qdt.ttinyint.ge(qdt.param('p1')).and(qdt.ttinyint.le(qdt.param('p2')));
}};

var q2 = {p1: 3, p2: 7, expected: [4, 5, 6], queryType: 3, ordered: false, predicate: function(qdt) {
  return qdt.ttinyint.gt(qdt.param('p1')).and(qdt.ttinyint.lt(qdt.param('p2')));
}};

var q3 = {p1: 3, p2: 7, expected: [7], queryType: 3, ordered: false, predicate: function(qdt) {
  return qdt.ttinyint.ne(qdt.param('p1')).and(qdt.ttinyint.eq(qdt.param('p2')));
}};

var q4 = {p1: 8, p2: 2, expected: [0, 1, 2, 8, 9], queryType: 3, ordered: false, predicate: function(qdt) {
  return qdt.ttinyint.ge(qdt.param('p1')).or(qdt.ttinyint.le(qdt.param('p2')));
}};

var q5 = {p1: 3, p2: 7, expected: [7, 8, 9], queryType: 3, ordered: false, predicate: function(qdt) {
  return qdt.ttinyint.gt(qdt.param('p1')).and(qdt.not((qdt.ttinyint.lt(qdt.param('p2')))));
}};

var q6 = {p1: 3, p2: 7, expected: [0, 1, 2, 3, 4, 5, 6, 8, 9], queryType: 3, ordered: false, predicate: function(qdt) {
  return qdt.not(qdt.ttinyint.ne(qdt.param('p1')).and(qdt.ttinyint.eq(qdt.param('p2'))));
}};

var q7 = {p1: 3, p2: 7, expected: [0, 1, 2, 3, 4, 5, 6, 8, 9], queryType: 3, ordered: false, predicate: function(qdt) {
  return (qdt.ttinyint.ne(qdt.param('p1')).and(qdt.ttinyint.eq(qdt.param('p2')))).not();
}};

var q8 = {p1: 3, p2: 7, expected: [4, 5, 6, 7], queryType: 3, ordered: false, predicate: function(qdt) {
  return qdt.ttinyint.gt(qdt.param('p1')).andNot((qdt.ttinyint.gt(qdt.param('p2'))));
}};

var q9 = {p1: 7, p2: 2, expected: [0, 1, 8, 9], queryType: 3, ordered: false, predicate: function(qdt) {
  return qdt.ttinyint.gt(qdt.param('p1')).orNot((qdt.ttinyint.ge(qdt.param('p2'))));
}};

var queryTests = [q1, q2, q3, q4, q5, q6, q7, q8, q9];

/***** Build and run queries ***/
var testQueries = new harness.ConcurrentTest("testQueries");
testQueries.run = function() {
  var testCase = this;
  var from = global.integraltypes;
  testCase.mappings = from;
  fail_openSession(testCase, function(session) {
      var i = 0, j = 0, completedTestCount = 0, testCount = queryTests.length;
      queryTests.forEach(function(queryTest) {
        // set up to check results
        queryTest.resultIds = [];
        ++i;
        queryTest.testName = 'q' + i;
        // each query test gets its own query domain type
        session.createQuery(from, function(err, q) {
          if (err) {
            testCase.appendErrorMessage('QueryTableScanTest.testQueries ' + queryTest.testName +
                ' returned error: ' + err);
            --testCount;
            return;
          }
          q.where(queryTest.predicate(q));
          testCase.errorIfNotEqual('Wrong query type for ' + queryTest.testName,
              queryTest.queryType, q.mynode_query_domain_type.queryType);
          q.execute(queryTest, function(err, results, queryTest) {
            testCase.errorIfNull("NoResults:q"+i+" ", results);
            if(results) {
              // check results
              // get the result ids in an array
              for (j = 0; j < results.length; ++j) {
                queryTest.resultIds[j] = results[j].id;
              }
              if (queryTest.expected.length !== results.length) {
                testCase.errorIfNotEqual('q' + i + ' wrong results: expected ' + 
                    queryTest.expected + '; actual: ' + queryTest.resultIds,
                    queryTest.expected.length, results.length);
              } else {
                if (!queryTest.ordered) {
                  // compare results without considering order
                  queryTest.resultIds.sort(function(a,b){return a-b;});
                }
                // compare the results one by one, in order
                for (j = 0; j < queryTest.expected.length; ++j) {
                  udebug.log_detail('QueryTableScanTest.testQueries ' + queryTest.testName + ' expected: '
                      + queryTest.expected[j] + ' actual: ' + queryTest.resultIds[j]);
                  testCase.errorIfNotEqual(queryTest.testName + ' wrong result at position ' + j,
                      queryTest.expected[j], queryTest.resultIds[j]);
                }
              }
            }
            if (++completedTestCount === testCount) {
              testCase.failOnError();
            }
          }, queryTest);
      });
    });
  });
};


module.exports.tests = [testQueries];

File: storage/ndb/clusterj/clusterj-test/src/main/java/testsuite/clusterj/QueryTableScanTest.java
/*
   Copyright (c) 2010, 2022, Oracle and/or its affiliates.
   Use is subject to license terms.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is also distributed with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have included with MySQL.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

package testsuite.clusterj;

import testsuite.clusterj.model.Employee;
import com.mysql.clusterj.query.QueryBuilder;
import com.mysql.clusterj.query.QueryDomainType;
import com.mysql.clusterj.query.Predicate;
import com.mysql.clusterj.query.PredicateOperand;
import com.mysql.clusterj.Query;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class QueryTableScanTest extends AbstractClusterJModelTest {

    @Override
    public void localSetUp() {
        createSessionFactory();
        session = sessionFactory.getSession();
        tx = session.currentTransaction();
        createEmployeeInstances(10);
        try {
            tx.begin();
            session.deletePersistentAll(Employee.class);
            tx.commit();
        } catch (Throwable t) {
            // ignore errors while deleting
        }
        tx.begin();
        session.makePersistentAll(employees);
        tx.commit();
        addTearDownClasses(Employee.class);
    }

    /** Test all queries using the same setup.
     * Fail if any errors during the tests.
     */
    public void testTableScan() {
        tableScanEqualQuery();
        tableScanGreaterEqualQuery();
        tableScanGreaterThanQuery();
        tableScanLessEqualQuery();
        tableScanLessThanQuery();
        failOnError();
    }
    public void tableScanEqualQuery() {

        tx.begin();
        // QueryBuilder is the sessionFactory for queries
        QueryBuilder builder = session.getQueryBuilder();
        // QueryDomainType is the main interface
        QueryDomainType dobj = builder.createQueryDefinition(Employee.class);

        // parameter name
        PredicateOperand param = dobj.param("name");
        // property name
        PredicateOperand column = dobj.get("name");
        // compare the column with the parameter
        Predicate compare = column.equal(param);
        // set the where clause into the query 
        dobj.where(compare);
        // create a query instance
        Query query = session.createQuery(dobj);

        // set the parameter value
        query.setParameter("name", "Employee number 8");
        // get the results
        List<Employee> results = query.getResultList();
        // consistency check the results
        consistencyCheck(results);
        // verify we got the right instance
        errorIfNotEqual("Wrong employee id returned from equal query: ",
                8, results.get(0).getId());
        tx.commit();
    }

    public void tableScanGreaterThanQuery() {

        tx.begin();
        // QueryBuilder is the sessionFactory for queries
        QueryBuilder builder = session.getQueryBuilder();
        // QueryDomainType is the main interface
        QueryDomainType dobj = builder.createQueryDefinition(Employee.class);

        // parameter name
        PredicateOperand param = dobj.param("name");
        // property name
        PredicateOperand column = dobj.get("name");
        // compare the column with the parameter
        Predicate compare = column.greaterThan(param);
        // set the where clause into the query 
        dobj.where(compare);
        // create a query instance
        Query query = session.createQuery(dobj);

        // set the parameter values
        query.setParameter("name", "Employee number 6");
        // get the results
        List<Employee> results = query.getResultList();
        // consistency check the results
        consistencyCheck(results);
        // verify we got the right instances
        Set<Integer> expected = new HashSet<Integer>();
        expected.add(7);
        expected.add(8);
        expected.add(9);
        Set<Integer> actual = new HashSet<Integer>();
        for (Employee emp: results) {
            actual.add(emp.getId());
        }
        errorIfNotEqual("Wrong employee ids returned from greaterThan query: ",
                expected, actual);
        tx.commit();
    }

    public void tableScanGreaterEqualQuery() {

        tx.begin();
        // QueryBuilder is the sessionFactory for queries
        QueryBuilder builder = session.getQueryBuilder();
        // QueryDomainType is the main interface
        QueryDomainType dobj = builder.createQueryDefinition(Employee.class);

        // parameter name
        PredicateOperand param = dobj.param("name");
        // property name
        PredicateOperand column = dobj.get("name");
        // compare the column with the parameter
        Predicate compare = column.greaterEqual(param);
        // set the where clause into the query 
        dobj.where(compare);
        // create a query instance
        Query query = session.createQuery(dobj);

        // set the parameter values
        query.setParameter("name", "Employee number 7");
        // get the results
        List<Employee> results = query.getResultList();
        // consistency check the results
        consistencyCheck(results);
        // verify we got the right instances
        Set<Integer> expected = new HashSet<Integer>();
        expected.add(7);
        expected.add(8);
        expected.add(9);
        Set<Integer> actual = new HashSet<Integer>();
        for (Employee emp: results) {
            actual.add(emp.getId());
        }
        errorIfNotEqual("Wrong employee ids returned from greaterEqual query: ",
                expected, actual);
        tx.commit();
    }

    public void tableScanLessThanQuery() {

        tx.begin();
        // QueryBuilder is the sessionFactory for queries
        QueryBuilder builder = session.getQueryBuilder();
        // QueryDomainType is the main interface
        QueryDomainType dobj = builder.createQueryDefinition(Employee.class);

        // parameter name
        PredicateOperand param = dobj.param("name");
        // property name
        PredicateOperand column = dobj.get("name");
        // compare the column with the parameter
        Predicate compare = column.lessThan(param);
        // set the where clause into the query 
        dobj.where(compare);
        // create a query instance
        Query query = session.createQuery(dobj);

        // set the parameter values
        query.setParameter("name", "Employee number 3");
        // get the results
        List<Employee> results = query.getResultList();
        // consistency check the results
        consistencyCheck(results);
        // verify we got the right instances
        Set<Integer> expected = new HashSet<Integer>();
        expected.add(0);
        expected.add(1);
        expected.add(2);
        Set<Integer> actual = new HashSet<Integer>();
        for (Employee emp: results) {
            actual.add(emp.getId());
        }
        errorIfNotEqual("Wrong employee ids returned from lessThan query: ",
                expected, actual);
        tx.commit();
    }

    public void tableScanLessEqualQuery() {

        tx.begin();
        // QueryBuilder is the sessionFactory for queries
        QueryBuilder builder = session.getQueryBuilder();
        // QueryDomainType is the main interface
        QueryDomainType dobj = builder.createQueryDefinition(Employee.class);

        // parameter name
        PredicateOperand param = dobj.param("name");
        // property name
        PredicateOperand column = dobj.get("name");
        // compare the column with the parameter
        Predicate compare = column.lessEqual(param);
        // set the where clause into the query 
        dobj.where(compare);
        // create a query instance
        Query query = session.createQuery(dobj);

        // set the parameter values
        query.setParameter("name", "Employee number 2");
        // get the results
        List<Employee> results = query.getResultList();
        // consistency check the results
        consistencyCheck(results);
        // verify we got the right instances
        Set<Integer> expected = new HashSet<Integer>();
        expected.add(0);
        expected.add(1);
        expected.add(2);
        Set<Integer> actual = new HashSet<Integer>();
        for (Employee emp: results) {
            actual.add(emp.getId());
        }
        errorIfNotEqual("Wrong employee ids returned from lessEqual query: ",
                expected, actual);
        tx.commit();
    }

}
